umbrella-feeds/src/lib.rs

```diff
@@ -0,0 +1,732 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! # Umbrella feeds
+//!
+//! Main contract for all on-chain data.
+//! Check `UmbrellaFeedsReader` to see how to integrate.
+//!
+//! ATTENTION: Keep the `upgradeNatively`/`unregister` entry points in this contract at all time and make sure their logic can be
+//! executed successfully via an invoke to the `atomicUpdate` entry point in the`registry` contract. Otherwise you will not be able to
+//! natively upgrade this contract via the `registry` contract anymore.
+//! Using the native upgradability mechanism for this contract is necessary to not break the `UmbrellaFeedsReader` contracts
+//! which include references to this `UmbrellaFeeds` contract.
+use concordium_std::*;
+use core::fmt::Debug;
+
+/// The name of this contract.
+const NAME: &str = "UmbrellaFeeds";
+
+/// The concept of a CHAIN_ID does not exist on Concordium but kept for consistency.
+const CHAIN_ID: u16 = 49228;
+
+#[derive(Serialize, SchemaType, Copy, Clone, Debug, PartialOrd, Ord, PartialEq, Eq)]
+pub struct PriceData {
+    /// This is a placeholder, that can be used for some additional data.
+    /// It is only used as marker for removed data (when data == u8::MAX) at the moment.
+    pub data: u8,
+    /// The heartbeat specifies how often the price data will be refreshed in case the price stays flat.
+    /// ATTENTION: u64 is used here instead of u24 (different from the original solidity smart contracts).
+    pub heartbeat: u64,
+    /// It is the time the validators run consensus to decide on the price data.
+    pub timestamp: Timestamp,
+    /// The price.
+    pub price: u128,
+}
+
+impl PriceData {
+    fn default() -> PriceData {
+        PriceData {
+            data: 0,
+            heartbeat: 0,
+            timestamp: Timestamp::from_timestamp_millis(0),
+            price: 0,
+        }
+    }
+}
+#[derive(Serial, DeserialWithState)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    /// Contract deployment time.
+    deployed_at: Timestamp,
+    /// Registry contract where the list of all addresses of this protocol is stored.
+    registry: ContractAddress,
+    /// StakingBank contract where list of validators is stored.
+    staking_bank: ContractAddress,
+    /// Minimal number of signatures required for accepting price submission (PoA).
+    required_signatures: u16,
+    /// Decimals for prices stored in this contract.
+    decimals: u8,
+    /// Map of all prices stored in this contract. It maps from the key to PriceData. The key for the map is the hash of the feed name.
+    /// E.g. for the "ETH-USD" feed, the key will be the hash("ETH-USD").
```

Comment:
https://github.com/Concordium/concordium-umbrella-oracle/pull/1#discussion_r1339807885

Reply:
I really don't think it has anything todo with security. Since no native support for strings exists in Solidity, hashes are used instead. Because umbrella is multi-chain protocol, it makes integration easier to have some consistency across chains in the way the smart contract logic works.

Reply:
Hashes have been replaced by the pure `String` now.